### ISSUE TITLE
contract-out fails

### ISSUE THREAD

**Issue by samth** [pycket/pycket#263]
Error message:

```
contract-out: not a provide sub-form in: (contract-out (make-constructor-style-printer (-> (-> any/c (or/c symbol? string?)) (-> any/c sequence?) (-> any/c output-port? (or/c #t #f 0 1) void?)))) ...
```

**Comment by samth** [comment]
This file fails with the same error:
```
#lang racket/base
(require racket/contract/base)
(define x 1)
(provide (contract-out [x any/c]))
```

**Comment by samth** [comment]
Here's a reduced example:

```
#lang racket/base
(module out racket/base
  (require (for-syntax racket/base 
                       racket/provide-transform))
  (provide contract-out)
  (define-syntax provide-contracted-vars 
    (make-provide-transformer
     (λ (stx modes) null)))

  (define-syntax contract-out
    (make-provide-pre-transformer
     (lambda (stx modes)
       #`(provide-contracted-vars)))))
(require 'out)
(provide (contract-out))
```

**Comment by cderici** [comment]
We know that we're making the `prop:provide-(pre)-transformer` properties three times, as opposed to Racket which is making it only twice (one for the `(#%require racket/private/reqprov)`, and the other for the `(for-syntax racket/provide-transform)`). (Recall that that property is defined at `provide-transform.rkt`)

On Pycket, right before the third evaluation I see a console_log saying : 

`[1543621447.960658] making instance : (submod 'tt out)`

So I guess it seems for some reason we're re-evaluating the required modules for the submodule.

**Comment by samth** [comment]
I've reduced the files to https://gist.github.com/0921cee87416ef9287a283dbf4311d43.

Note that you if make the requires of racket/private/reqprov2 and racket/provide-transform2 just relative paths, the error goes away.

**Comment by samth** [comment]
Also, `module-declared?` with a `#t` second argument does require things. That shouldn't happen because we got an error, but the error is currently swallowed by `call_interpret`.

**Comment by samth** [comment]
Here's a further reduced example: https://gist.github.com/83a37b94b11510323ab77003b8f9bf93

I think I've identified the problem. When you `require` the `racket/provide-transform2` module in two _different_ ways (ie, one by `racket/provide-transform2` and one by `"../provide-transform2.rkt"`) then we see the problem. If we require them both the _same_ way (ie, both `racket/provide-transform2`) then the problem goes away.

This means that the expander can't tell that we already have an instance for the module, and so it creates another one. My guess is that this is a hash table bug. There might be a hash table keyed on `resolved-module-path` or on `module-path-index` both of which have `prop:equal+hash` which we might not be treating correctly.

**Comment by samth** [comment]
The good news is that this means that we're creating way too many instances as it is, and so things should get faster when this is fixed.

**Comment by samth** [comment]
This was fixed by the `simplify-path` change, a76a5a3b3ab4d2cc.
